### PR TITLE
[Feature] Add reasoning_eos_policy to close think on mid-reasoning EOS

### LINKED ISSUE #55420: [Feature]: Option to suppress EOS inside the reasoning block and force `reasoning_end` first (Qwen3.8 samples `<|im_end|>` mid-`<think>`, returning empty `content` with `finish_reason: stop`)
### 🚀 The feature, motivation and pitch

**Problem.** With `--reasoning-parser qwen3`, Qwen3.8-27B sometimes samples `<|im_end|>` *inside* the `<think>` block. vLLM treats it as EOS, the request ends with `finish_reason: "stop"`, the parser routes everything to `reasoning_content`, and `content` is empty. The call is HTTP 200 and looks complete; thousands of tokens were spent; the budget was not reached. A client cannot tell this apart from a legitimate empty answer.

This is not a parser bug and not truncation. I made the stop token visible with `logprobs` on `/v1/completions` (prompt = chat-templated token ids, `thinking_token_budget: 16384`, `max_tokens: 32768`, shipped sampling defaults):

| step before stop | sampled | `<|im_end|>` logprob |
|---|---|---|
| −3 | ` Ma` | −8.5 |
| −2 | `\n` | −5.8 |
| −1 | `3` | −6.3 |
| **stop** | **`<|im_end|>`** | **−0.32** (alternative `7` at −1.3) |

EOS sits in the top-2 at −3…−8 for thousands of steps and then wins at one step. It is a per-token hazard, so stops scatter between 4k and 16k generated tokens. It is content-triggered: it appears during long, near-identical enumeration in the reasoning (e.g. tallying hundreds of `00433: Geschichte` lines); prose-style reasoning ran to 65k tokens without it.

**It is the model, not a serve setting.** Same 32.7k-token prompt, sequential runs, "stop" = EOS inside `<think>`:

| variable | result |
|---|---|
| MTP (n=3) on / off | 12/12 and 9/13 stop |
| 4 concurrent / sequential | stops either way |
| GDN state bf16 / fp32 | stops either way |
| KV cache fp8_e4m3 / bf16 | stops either way |
| NVFP4+FP8 weights / BF16 weights / base `unsloth/Qwen3.8-27B-NVFP4` | all stop |
| vLLM 0.27.1 (sm_120 build) / **upstream llama.cpp master, Q8_0 GGUF** | both stop, same logprob signature |

Independent prior reports of the same symptom: QwenLM/Qwen3.8#216 (~1,000 measured calls: `reasoning_effort: xhigh` fails 19–38 %, `temperature: 0` fails 12/12, on both vLLM and llama.cpp; the only mitigation found there is `frequency_penalty: 0.3`), sgl-project/sglang#24839 (same failure on Qwen3.6-27B-FP8 with the qwen3 parser; closed without an engine change). Related vLLM issues that are downstream symptoms of the same event: #17357 and #48645 (reply without `</think>` lands in `reasoning_content`), #39056 (tool call emitted inside `<think>`).

**Pitch.** vLLM already has the machinery to fix this at the right layer. `ThinkingBudgetStateHolder` (`vllm/v1/sample/thinking_budget_state.py`) tracks `in_think` / `in_end` per request, knows the `reasoning_end` token ids from the reasoning parser, handles speculative tokens (`in_spec_mode`, `force_index`), and already *forces* the end tokens when the budget is exhausted. The proposal is to let the same holder react to an **EOS that would end generation while `in_think` is true**, exactly as it reacts to budget exhaustion:

- new per-request sampling parameter, e.g. `reasoning_eos_policy: "stop" | "force_end"` (default `"stop"`, today's behavior), exposed on Chat Completions / Completions like `thinking_token_budget`;
- with `"force_end"`, when the request is `in_think` and the sampled token (or an accepted speculative token) is an EOS/stop token, mask EOS in the logits for that step and set `force_index` so `reasoning_end_str` is emitted; generation then continues into the answer;
- the reasoning parser is untouched, `ignore_eos` keeps its meaning (this policy never suppresses EOS outside the think block), and the tool-call implicit end discussed in #44676 must count as leaving the think block so a `<tool_call>` inside `<think>` is not "protected".

Expected effect on the runs above: the model closes its reasoning and answers instead of returning nothing. On the Qwen3.8 issue's numbers this would turn a 19–38 % silent-failure rate at the model's default effort into a normal answer path, and it is engine-side so every OpenAI-only client benefits without knowing about it.

### Alternatives

- **Client retry on `stop` + empty `content`**: works, but doubles cost on a call that already spent 8–16k tokens, and it is per-client.
- **`frequency_penalty: 0.3`** (QwenLM/Qwen3.8#216): changes the whole distribution to dodge one token; not a fix.
- **`logit_bias` on `<|im_end|>`**: blocks the legitimate end of the answer too, and `logit_bias` does not work with speculative decoding.
- **`ignore_eos`**: never stops; not usable.
- **A larger `thinking_token_budget`**: irrelevant, the stop precedes the budget.
- **Doing it in the reasoning parser / output processor**: too late, the engine has already finished the request; the sampler is the only place that can substitute `</think>` for EOS.

### Additional context

- Environment for the measurements: RTX PRO 6000 Blackwell (sm_120), vLLM 0.27.1 (AEON sm_120 build; `ThinkingBudgetStateHolder` present), `--reasoning-parser qwen3 --kv-cache-dtype fp8_e4m3 --attention-backend TRITON_ATTN`, and upstream llama.cpp `427291b5b` for the cross-engine check. Full method, logs and the per-run tables are in my workstation diary; I can attach the raw logprob dumps.
- I am happy to draft the PR if the maintainers agree with the placement in `ThinkingBudgetStateHolder` and the parameter shape; the recent changes there (#45984, #46662, #49613) suggest it is the maintained home for reasoning-boundary forcing.
- Related to the request in #55382 (effort → budget map): both are about letting the engine, rather than every client, keep reasoning models from failing silently.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by ayush-singh-0601**
Fixes #55420

## Purpose

Qwen3.8 (and similar reasoning models) sometimes sample `<|im_end|>` while still inside `<think>`. vLLM treats that as a real stop, the parser parks the unfinished think block in `reasoning_content`, and `content` is empty with `finish_reason: "stop"`. HTTP 200, so a client cannot tell this apart from a legitimate empty answer.

This adds a per-request sampling parameter, same placement as `thinking_token_budget`:

- `reasoning_eos_policy="stop"` (default): current behavior
- `reasoning_eos_policy="force_end"`: while the request is in-think, mask EOS/stop tokens. If EOS would be sampled (greedy-top, or an accepted speculative token), force `reasoning_end_str` on the existing `ThinkingBudgetStateHolder` / v2 thinking-budget kernel path, then continue into the answer

`ignore_eos` is unchanged. The policy does nothing outside a think block. A Qwen3 `<tool_call>` started inside `<think>` counts as leaving think, so tool-call tokens are not rewritten (the implicit-end case from #44676).

I searched open PRs for `55420`, `reasoning_eos_policy`, and `force_end`. I did not find another PR implementing this parameter. Closest related work is stop-string handling inside reasoning (#52486) and loop-breaking by forcing reasoning end (#52677); those are different problems.

AI assistance was used to implement and test this. I reviewed the sampler/holder changes and the API wiring.

## Test Plan

```bash
.venv/bin/ruff check <changed python files>
.venv/bin/ruff format --check <changed python files>
PYTHONPATH=. .venv/bin/python -m pytest \
  tests/v1/sample/test_thinking_budget_state.py \
  tests/entrypoints/openai/chat_completion/test_thinking_token_budget_validation.py \
  -v --noconftest
```

v2 GPU kernel tests in `tests/v1/worker/test_gpu_thinking_budget.py` skip without CUDA. I could not run a Qwen3.8 serve eval in this environment.

## Test Result

- ruff check/format: passed on the changed Python files
- 29 unit tests passed (holder behavior, OpenAI Chat/Completions validation and `to_sampling_params`)
- GPU kernel tests: not run (no CUDA)
- Model eval: not run. I do not have a Qwen3.8 checkpoint or GPU here. Happy to follow up with a serve log if someone can try `reasoning_eos_policy=force_end` on the reproduction from the issue

## Docs

`docs/features/reasoning_outputs.md` has a short section and curl/offline examples.

### CODE REVIEW

File: tests/entrypoints/openai/chat_completion/test_thinking_token_budget_validation.py [hunk 1]
```diff
@@ -65,6 +65,64 @@ def test_completion_request_accepts_valid_thinking_token_budget():
...
+            "model": "qwen",
+            "messages": [{"role": "user", "content": "hello"}],
+            "reasoning_eos_policy": "force_end",
+        }
+    )
+    assert request.reasoning_eos_policy == "force_end"
+
+
+def test_chat_completion_request_rejects_invalid_reasoning_eos_policy():
+    with pytest.raises(Exception, match="reasoning_eos_policy"):
```
[the remark this answers is not in the document]
Comment (ayush-singh-0601) [reply tests/entrypoints/openai/chat_completion/test_thinking_token_budget_validation.py]: Addressed in d91bb5c. Both invalid-policy tests now use `pytest.raises(VLLMValidationError, match="reasoning_eos_policy")` so an unrelated exception cannot pass them.

File: vllm/sampling_params.py [hunk 2]
```diff
@@ -393,6 +428,7 @@ def from_optional(
         stop_token_ids: list[int] | None = None,
         bad_words: list[str] | None = None,
         thinking_token_budget: int | None = None,
+        reasoning_eos_policy: Literal["stop", "force_end"] = "stop",
```
[the remark this answers is not in the document]
Comment (ayush-singh-0601) [reply vllm/sampling_params.py]: Addressed in d91bb5c. `reasoning_eos_policy` is now the last parameter of `SamplingParams.from_optional`, so positional arguments after `thinking_token_budget` still bind to `include_stop_str_in_output`. Added `test_from_optional_preserves_positional_include_stop_str_in_output`.

File: vllm/v1/sample/thinking_budget_state.py [hunk 3]
```diff
@@ -227,8 +261,47 @@ def _init_state_entry(
...
+        stop_ids = set(state.get("stop_token_ids") or [])
+        if not stop_ids:
+            return
+        for i, token_id in enumerate(state.get("spec_token_ids") or []):
+            if token_id in stop_ids:
+                state["in_think"] = False
+                state["in_end"] = True
+                state["end_count"] = 0
+                state["force_index"] = [i]
+                return
```
[the remark this answers is not in the document]
Comment (ayush-singh-0601) [reply vllm/v1/sample/thinking_budget_state.py]: Addressed in d91bb5c. Spec tokens are scanned in order and force-end stops at the first implicit or natural reasoning exit. `[TOOL_CALL, EOS]` no longer rewrites the EOS. Coverage is in `test_force_end_does_not_rewrite_eos_after_speculative_tool_call`.

File: vllm/v1/sample/thinking_budget_state.py [hunk 4]
```diff
@@ -502,6 +587,111 @@ def _apply_forcing_to_logits(
...
+    def _request_row_count(
+        self,
+        seq_idx: int,
+        predict_bonus_token: bool,
+        spec_token_ids_for_layout: list[list[int]],
+    ) -> int:
+        if not self.in_spec_mode or predict_bonus_token:
+            return 1
+        if seq_idx < len(spec_token_ids_for_layout):
+            return max(len(spec_token_ids_for_layout[seq_idx]), 1)
```
[the remark this answers is not in the document]
Comment (ayush-singh-0601) [reply vllm/v1/sample/thinking_budget_state.py]: Addressed in d91bb5c. When `predict_bonus_token` is false, the row count is the actual spec length, including zero, so an empty first result does not consume the next request's logits row. Added `test_force_end_empty_spec_does_not_mask_next_request_row`.

File: vllm/v1/worker/gpu/sample/thinking_budget.py [hunk 5]
```diff
@@ -77,20 +86,50 @@ def __init__(
...
         if self.thinking_token_budget.np[req_idx] != budget:
             self.thinking_token_budget.np[req_idx] = budget
             self._budget_dirty = True
+        if self._policy_cpu[req_idx] != policy:
+            self._policy_cpu[req_idx] = policy
+            self._eos_dirty = True
+        self._stop_cpu[req_idx].fill(-1)
+        if policy == 1:
+            stop_ids = sampling_params.stop_token_ids_that_finish_request()
+            n_stop = min(len(stop_ids), _MAX_STOP_IDS)
```
[the remark this answers is not in the document]
Comment (ayush-singh-0601) [reply vllm/v1/worker/gpu/sample/thinking_budget.py]: Addressed in d91bb5c. force_end now raises if there are more finishing stop token IDs than the GPU table (16) instead of silently truncating. GPU test: test_v2_force_end_rejects_too_many_stop_token_ids.

File: vllm/v1/worker/gpu/sample/thinking_budget.py [hunk 6]
```diff
@@ -221,6 +272,21 @@ def _update_committed_marker_cache_kernel(
...
+                        all_token_ids_ptr
+                        + req_state_idx * all_token_ids_stride
+                        + offs
+                        + j,
+                        mask=offs + j < total_len,
+                        other=-1,
+                    )
+                    impl_match = impl_match & (actual == expected)
+                last_impl = tl.max(tl.where(impl_match, offs, -1), axis=0)
+                last_end = tl.maximum(last_end, last_impl)
```
[the remark this answers is not in the document]
Comment (ayush-singh-0601) [reply vllm/v1/worker/gpu/sample/thinking_budget.py:296]: Addressed in d91bb5c. Cold-scan now keeps the max start/end across blocks with `tl.maximum`, so an implicit end is not overwritten by a later block that only has `<think>`. Regression: `test_v2_force_end_keeps_tool_call_across_cold_scan_blocks`.

**Comment by ayush-singh-0601** [comment]
Hi, the implementation and requested fixes are ready. Could someone please review/verify the PR and add the appropriate ready/verified label so CI can run? Thanks!

**Comment by ayush-singh-0601** [comment]
Rebased onto the latest `main` and resolved the adapter conflict, keeping both the upstream parser updates and this PR's implicit reasoning-end handling. The branch is conflict-free now, and pre-commit passes.

**Comment by ayush-singh-0601** [comment]
Rebased again onto the latest `main` (`ae71862`) and updated the positional-compatibility regression test for the upstream `watermarking` parameter added to `SamplingParams.from_optional`.

Local validation after the rebase:

- targeted pytest suite: **32 passed**
- pre-commit on all changed files: **all applicable hooks passed** (including Ruff, mypy, Markdown, SPDX/config checks, and Rust formatting)

All review threads are resolved and GitHub reports the PR as mergeable. Could a maintainer please review and apply the appropriate ready/verified state so the full CI can run?
